yarp/yarp_compiler.c

```diff
@@ -970,13 +970,14 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
       case YP_NODE_PROGRAM_NODE: {
           yp_program_node_t *program_node = (yp_program_node_t *) node;
 
+          yp_scope_node_t scope_node;
+          yp_scope_node_init((yp_node_t *)node, &scope_node);
           if (program_node->statements->body.size == 0) {
               ADD_INSN(ret, &dummy_line_node, putnil);
           } else {
-              yp_compile_node(iseq, (yp_node_t *) program_node->statements, ret, src, popped, compile_context);
+              yp_compile_node(iseq, (yp_node_t *) &scope_node, ret, src, popped, compile_context);
```

Comment:
I believe this is a strict-aliasing issue in C. I think the solution is:

```suggestion
              yp_scope_node_t *node = &scope_node;
              yp_compile_node(iseq, (yp_node_t *) node, ret, src, popped, compile_context);
```

Truthfully I still don't truly understand strict aliasing, but this is what I've done in the past that @XrXr has pointed me to. Alan can you confirm or deny what I'm saying here? I'm out of my depth.

Reply:
It might be fine here for strict aliasing because there is actually a `yp_node_t` inside `scope_node_t`. But you could get rid of any casting by passing `&scope_node->base`.